### ISSUE TITLE
Bitwise ~ operator

### ISSUE THREAD

**Issue by mstaehely** [kelloggm/checker-framework#104]
The following code issues an incompatible return warning:
```
public static @NonNegative int bitwiseTest() {
    String str = "This is a string.";
    int idx = str.indexOf('z');
    if (idx < 0) {
        idx = ~idx;
    }
    return idx;
}
```
I'm not entirely sure about whether or not this is an issue; when 0 has the ~ operator applied to it, it will be -1. Other than that, every value from -1 to Integer.MIN_VAL should become positive. Is this even within the scope of the checker?

**Comment by mernst** [comment]
Here is a simplified test case, with explanations:

```java
public class BitwiseTest {

  public static @NonNegative int bitwiseTest(@GTENegativeOne idx) {
    if (idx < 0) {
      // idx is negative; more specifically, it is -1
      idx = ~idx;
      // idx is non-negative
    }
    // idx is non-negative
    return idx;
  }

}
```

The Index Checker does not have a way of expressing "`idx` is negative" or "`idx' is -1", which are required in order to conclude that the final result is non-negative.  Doing so would require integration with the Constant Value Checker or the upcoming Int Range Checker.

Does this come up in practice?

**Comment by mstaehely** [comment]
Yes, although infrequently. It came up primarily when scanning or parsing Strings, where the position of a found value would be returned and, if not found then ~position would be returned to indicate where that value should have been. 

It also came up when working with Arrays.binarySearch. Here is a code snippet:
```
public int getOffset(long instant) {
    long[] transitions = iTransitions; // iTransitions is an array of longs within this method's class
    int idx = Arrays.binarySearch(transitions, instant);
    if (idx >= 0) {
        return iWallOffsets[idx] // another array within this method's class
    }
    idx = ~idx
    // more code where idx might be used to index
}
```

If the @NonNegative return annotation is removed but the index argument is @NonNegative, or listed as ```@IndexFor("iWallOffsets")``` it will still end up issuing an incompatible assignment warning, which makes sense as binarySearch can return a negative value. This value never gets used if it's negative. Rather, the ~ operator is applied. However, since the checker does not have a way to express the negative or -1 states as you noted, I'm seeing how this warning would be difficult to avoid.

**Comment by mernst** [comment]
Interesting.  I hadn't seen that code idiom before.  It's too complex for the Index Checker for now, so chalk it up to Index Checker imprecision.

**Comment by mernst** [comment]
Closed in favor of #124, which has a concrete example and a proposal for fixing.
